pkg/services/govmomi/constants.go

```diff
@@ -23,3 +23,12 @@ const (
 const (
 	guestInfoKeyMetadata = "guestinfo.metadata"
 )
+
+const (
+	// AnnotationGuestShutdown indicates whether the machine should perform a guest shutdown
+	// if VMware Tools is available.
+	AnnotationGuestShutdown = "vsphere.infrastructure.cluster.x-k8s.io/guest-shutdown"
+
+	// AnnotationGuestShutdownTimeout sets the timeout for guest shutdown.
+	AnnotationGuestShutdownTimeout = "vsphere.infrastructure.cluster.x-k8s.io/guest-shutdown-timeout"
```

Comment:
Is there a reason to use an annotation as opposed to spec field?

Sorry if this has been discussed before.

cc @srm09 @yastij

Reply:
I believe `spec` field and `annotations` would do the same to implement this feature.
I prefer not to make changes to the CRDs frequently to control this kind of subtle behavior.
(P.S I haven't verified if the annotations could be set yet)

For most cases I don't think the user would change these parameters for these two considerations

1. It's required to have VMware tools installed in the image.
Guest shutdown is recommended and should work for most cases.
The `guest-shutdown` flag is only added for consistency when the user really wants the old behavior back.

2. Forcibly powering off VM would only kick in when guest shutdown stalls.
This would be more like a safeguard and should not happen frequently in a cluster.
We should investigate why it stalls when this happens.

btw, I'm not sure the default 10m is enough to drain a node and perform graceful shutdown when there are still pods running.
Maybe we should set it to longer time to be safe.

Reply:
> I believe spec field and annotations would do the same to implement this feature.

Yes, but an annotation is essentially a hidden API. I do expect users to want to change this depending on the specifics of their application in terms of how long it takes to shutdown - a clustered DB might need some time to terminate, for example. Given there's potential for data loss, I would prefer making the API explicit using a new spec field and a defaulted value that is visible when looking at the CRD.

> It's required to have VMware tools installed in the image.

VMware tools is pretty much a requirement and baked into Image Builder.

> btw, I'm not sure the default 10m is enough to drain a node and perform graceful shutdown when there are still pods running.

Yes, this should be increased to a "safe" value.

The release note needs to be expanded on to make the behaviour clear to users.

Reply:
Annotations are not APIs, and when they are changed, it is still a breaking behavior. If a user is upset that a behavior changes, it matters not if it is an annotation or API, so make it an API.

Reply:
@akutz kindly pointed me to the docs for VM Operator, and the timeout is currently set to 5 minutes. I suggest we set the default to 5 minutes, and have a spec field to make the timeout configurable. We should closely match https://vm-operator.readthedocs.io/en/latest/concepts/workloads/vm/#power-op-mode so people can move between govmomi and VM operator modes as easily as possible.

I'm going to open an issue in VM Operator to also make the timeout configurable.

Reply:
https://github.com/vmware-tanzu/vm-operator/issues/179

Reply:
In that case I would like to propose these changes to both CRDs `VSphereMachineSpec` and `VSphereVMSpec` as alternative to the annotations.

- `PowerOffMode`: same as `spec.powerOffMode` in VMOperator#VirtualMachine. Defaults to `hard` to keep the old behavior. P.S: I'm wondering if we should change the default to `trySoft` even it's a behavior change to previous release.
- `GuestShutdownTimeout`: The timeout value before waiting for guest shutdown completion when `PowerOffMode` is set to `trySoft`. After the timeout the VM will be powered off forcibly. The default timeout is 1h.